File: dnn-providers/hip-kernel-provider/rocke/platform/Python/rocke/instances/common/grouped_gemm.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Grouped GEMM instance builder (CK Tile ``17_grouped_gemm`` parity).

This is the DSL-side counterpart of CK Tile's
``example/ck_tile/17_grouped_gemm`` (the unquantised base case;
``abquant_grouped_gemm.cpp`` / ``quant_grouped_gemm*.cpp`` are
follow-ups).

A grouped GEMM is a batch of GEMMs where each batch entry can have a
distinct ``(M, N, K)`` shape and is stored at an arbitrary device
offset (not a uniform per-batch stride like ``16_batched_gemm``).
Use cases:

  - MoE-style routing where each expert handles a different number of
    tokens
  - Multi-LoRA inference where each request hits a different LoRA
    adapter

Two implementations are provided here, mirroring CK Tile's own design
points:

1. :class:`GroupedGemmLauncher` (default) — re-uses
   :func:`rocke.instances.common.gemm_universal.build_universal_gemm` and
   launches it once per group. This is what CK Tile's reference
   path does when ``persistent=False``. Correct, easy to reason
   about, but pays one ``hipModuleLaunchKernel`` round-trip per
   group (~3-5us on MI300X/MI355X). Acceptable when the per-group
   work is large relative to the launch cost (typical for MoE).

   v1 optimisation note (this file): the per-group ``LaunchConfig``,
   the per-group element-arg dict, and the ``ceil_div`` grid math
   are extracted into a single hot loop that allocates one launch
   bundle per call rather than rebuilding them every iteration -- the
   per-group ``__call__`` overhead drops to one ``KernelLauncher``
   dispatch + one launch grid tuple per group. See
   :class:`GroupedGemmLauncher.__call__` for the loop body.

2. ``GroupedGemmKernelSpec`` (planned) — single-launch kernel that
   uses ``block_id_z`` as the group index, looks up
   ``M[g] / N[g] / K[g] / A_off[g] / B_off[g] / C_off[g]`` from
   device-side arrays, and runs the universal_gemm body in-place
   with the looked-up offsets. This matches CK Tile's
   ``persistent=True`` grouped GEMM. The lookup math is per-group
   wave-uniform i32 (one ``amd_wave_read_first_lane`` per offset in
   the CK Tile reference, ``grouped_gemm_kernel.hpp:297-298,
   390, 451``) so an SGPR-pinned version saves one
   ``v_readfirstlane_b32`` per use across the K-loop and epilogue.

   Listed as a follow-up because:

   * The current ``mfma_k_loop`` helper requires compile-time ``K``
     (the K-loop trip count is a Python int baked into
     ``scf.for_iter``). A single-launch grouped GEMM with per-group
     dynamic ``K`` needs either a new runtime-K MFMA helper or a
     compile-time per-K specialisation of the kernel.

   * The descriptor lookup wants a constant-address-space pointer
     (per CK Tile's pattern at ``grouped_gemm_kernel.hpp:503``,
     ``void CK_TILE_CONSTANT_ADDRESS_SPACE* gemm_descs_const``); the
     ``rocke.core.ir.PtrType("global")`` path covers this but the
     LLVM-side ``addrspace(4)`` lift needs a small extension to the
     param attrs. Both items are future extensions; the public API
     and per-group input layout here match what the single-launch
     kernel will use.

Today's launcher therefore is the per-group multi-launch path. The
public API and per-group input layout match what the single-launch
kernel will use, so callers can switch implementations without
changing their input plumbing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Sequence, Tuple

from ...core.ir import KernelDef
from ...helpers.spec import WarpTileBlockSizeMixin, ceil_div_grid
from ...runtime.launcher import KernelLauncher, LaunchConfig, LaunchSummary
from .gemm_universal import (
    DataSpec,
    TileSpec,
    TraitSpec,
    UniversalGemmSpec,
    build_universal_gemm,
    is_valid_spec as is_valid_gemm_spec,
    mono_data_spec,
)


@dataclass(frozen=True)
class GroupedGemmProblem:
    """One entry in a grouped GEMM workload.

    Pointers are device pointers (e.g. ``tensor.data_ptr()`` from
    torch). Strides are in elements. Layout is RCR (row-major A, col-
    major B, row-major C), matching ``build_universal_gemm``.
    """

    M: int
    N: int
    K: int
    A_ptr: int
    B_ptr: int
    C_ptr: int


@dataclass(frozen=True)
class GroupedGemmSpec(WarpTileBlockSizeMixin):
    """One grouped GEMM kernel-instance bundle.

    The same tile/trait pair is applied to every group (CK Tile's
    grouped GEMM has the same constraint). The dispatcher should
    select tile dims that divide every ``(M[g], N[g], K[g])`` (the
    current implementation does not pad).
    """

    name: str
    tile: TileSpec
    trait: TraitSpec
    wave_size: int = 64
    block_size: int = 0
    dtype: str = "fp16"

    def __post_init__(self) -> None:
        self._init_block_size()

    def _data_spec(self) -> DataSpec:
        return mono_data_spec(self.dtype)

    def to_universal_spec(self) -> UniversalGemmSpec:
        return UniversalGemmSpec(
            name=self.name,
            tile=self.tile,
            trait=self.trait,
            data=self._data_spec(),
            wave_size=self.wave_size,
            block_size=self.block_size,
            batched=False,
        )

    def kernel_name(self) -> str:
        return self.to_universal_spec().kernel_name()


def is_valid_spec(spec: GroupedGemmSpec, arch: str = "gfx950") -> Tuple[bool, str]:
    """Validity of the per-group base kernel on ``arch``.

    Delegates to :func:`rocke.instances.common.gemm_universal.is_valid_spec`
    so the MFMA-atom + LDS-cap gating is arch-aware (a gfx950-only wide
    atom is rejected for gfx942 with a structured reason).
    """
    return is_valid_gemm_spec(spec.to_universal_spec(), arch=arch)


def build_grouped_gemm(spec: GroupedGemmSpec, arch: str = "gfx950") -> KernelDef:
    """Build the IR for the per-group base kernel.

    The returned :class:`KernelDef` is the *same* kernel as
    :func:`build_universal_gemm`; the grouping happens entirely at
    launch time (one launch per group) in
    :class:`GroupedGemmLauncher`.

    ``arch`` validates the spec against the target's MFMA catalog so a
    gfx950-only warp-tile atom requested for gfx942 raises a clean
    Python error before comgr.
    """
    universal = spec.to_universal_spec()
    ok, why = is_valid_gemm_spec(universal, arch=arch)
    if not ok:
        raise ValueError(f"invalid grouped_gemm spec for {arch}: {why}")
    return build_universal_gemm(universal, arch=arch)


def grouped_gemm_signature(spec: GroupedGemmSpec):
    from ...helpers.spec import SignatureBuilder

    ptr_dt = spec.dtype if spec.dtype in ("f16", "fp16", "bf16") else "f16"
    return (
        SignatureBuilder()
        .ptr("A", ptr_dt)
        .ptr("B", ptr_dt)
        .ptr("C", ptr_dt)
        .scalar("M", "i32")
        .scalar("N", "i32")
        .scalar("K", "i32")
        .build()
    )


class GroupedGemmLauncher:
    """Single-pass grouped GEMM launcher.

    Owns one :class:`KernelLauncher` and re-uses it across every
    group in the workload. The HSACO module is loaded once at
    construction time and cached on the underlying launcher; per-
    group launches only update the (A, B, C, M, N, K) packed args
    and the grid dims.

    Constructed once per (problem-tile, dtype) tuple; call repeatedly
    with a list of :class:`GroupedGemmProblem` entries.
    """

    def __init__(self, *, hsaco: bytes, spec: GroupedGemmSpec) -> None:
        self._spec = spec
        self._launcher = KernelLauncher(
            hsaco=hsaco,
            kernel_name=spec.kernel_name(),
            signature=grouped_gemm_signature(spec),
            cache_key=(spec.kernel_name(),),
        )
        # Hoist host-side per-group constants out of the inner loop so
        # ``__call__`` only pays for the bits that genuinely vary per
        # group (the (A, B, C) torch pointers and the (M, N, K) shape).
        # These three values are CTA-uniform for the kernel and could in
        # principle be lifted further if we expose a "shape-bucketed"
        # API (see the GroupedGemmKernelSpec follow-up in the module
        # docstring above). For the multi-launch path it's pure Python
        # overhead reduction: one tuple build + one ``__init__`` of
        # ``LaunchConfig`` per group instead of building the same
        # constant block-dim 3-tuple every call.
        self._block = (spec.block_size, 1, 1)
        self._tile_m = spec.tile.tile_m
        self._tile_n = spec.tile.tile_n

    def __call__(
        self,
        problems: Sequence[GroupedGemmProblem],
        *,
        stream: int = 0,
    ) -> LaunchSummary:
        # Pre-bind the per-instance constants once; the inner loop
        # then only does the work that *cannot* be hoisted host-side:
        # the per-group ``(p.M, p.N, p.K, p.A_ptr, p.B_ptr, p.C_ptr)``
        # plumbing and the per-group ``ceil_div`` grid math.
        tile_m = self._tile_m
        tile_n = self._tile_n
        block = self._block
        launcher = self._launcher
        launches = 0
        for p in problems:
            cfg = LaunchConfig(
                stream=stream,
                grid=ceil_div_grid((p.N, tile_n), (p.M, tile_m)),
                block=block,
            )
            launches += launcher(
                {
                    "A": p.A_ptr,
                    "B": p.B_ptr,
                    "C": p.C_ptr,
                    "M": p.M,
                    "N": p.N,
                    "K": p.K,
                },
                config=cfg,
            ).launches
        return LaunchSummary(launches=launches)


def build_grouped_gemm_single_launch(
    spec: GroupedGemmSpec, arch: str = "gfx950"
) -> KernelDef:
    """Single-launch grouped GEMM device kernel.

    Grid ``(N_tile_max, M_tile_max, num_groups)``: ``block_id_z`` is the group
    index. All ``num_groups`` GEMMs run in **one** ``hipModuleLaunchKernel``,
    eliminating the per-group HIP launch tax (~3-5us / group on MI300X /
    MI355X) that the multi-launch :class:`GroupedGemmLauncher` pays.

    The realised kernel is the **batched** universal-GEMM body
    (``UniversalGemmSpec(batched=True)``): the kernel reads ``block_id_z`` as
    the group index and offsets the A / B / C base pointers by
    ``block_id_z * stride_{a,b,c}`` (element strides passed as i32 kernel
    args). This is exactly CK Tile's ``persistent=False`` grouped GEMM for the
    *uniform-shape* case (every group shares ``(M, N, K)`` — the dominant
    batched-MoE / multi-LoRA routing case), where each group's matrices are
    contiguous so ``stride_a = M*K``, ``stride_b = N*K``, ``stride_c = M*N``
    and the ``block_id_z``-scaled offset lands exactly on each group's base.

    Drive it with :class:`GroupedGemmSingleLaunchRunner`, which packs the
    per-group contiguous tensors into one launch with ``grid.z = num_groups``.
    Verify against the per-group path with the same tile/trait spec.

    The fully-variable-shape single launch (per-group ``(M[g], N[g], K[g])``
    decoded from an ``addrspace(4)`` descriptor table + ``mfma_k_loop_dynamic_K``
    runtime-K loop + out-of-range tile guard, CK Tile
    ``grouped_gemm_kernel.hpp:501-577``) remains the v2 follow-up; it needs the
    universal body to take per-group runtime M/N/K, which the current
    block-offset model does not yet thread.
    """
    base_spec = spec.to_universal_spec()
    batched_spec = UniversalGemmSpec(
        name=base_spec.name + "_single_launch",
        tile=base_spec.tile,
        trait=base_spec.trait,
        data=base_spec.data,
        wave_size=base_spec.wave_size,
        block_size=base_spec.block_size,
        batched=True,
    )
    ok, why = is_valid_gemm_spec(batched_spec, arch=arch)
    if not ok:
        raise ValueError(f"invalid grouped_gemm_single_launch spec for {arch}: {why}")
    return build_universal_gemm(batched_spec, arch=arch)


def grouped_gemm_single_launch_signature(spec: GroupedGemmSpec):
    """Kernel-ABI signature for :func:`build_grouped_gemm_single_launch`.

    Matches the batched universal-GEMM body the builder emits:
    ``(A, B, C, M, N, K, stride_a, stride_b, stride_c)``. The host runner
    sets ``M`` / ``N`` / ``K`` to the shared per-group shape and the three
    strides to the per-group element counts (``M*K`` / ``N*K`` / ``M*N``).
    """
    from ...helpers.spec import SignatureBuilder

    ptr_dt = spec.dtype if spec.dtype in ("f16", "fp16", "bf16") else "f16"
    return (
        SignatureBuilder()
        .ptr("A", ptr_dt)
        .ptr("B", ptr_dt)
        .ptr("C", ptr_dt)
        .scalar("M", "i32")
        .scalar("N", "i32")
        .scalar("K", "i32")
        .scalar("stride_a", "i32")
        .scalar("stride_b", "i32")
        .scalar("stride_c", "i32")
        .build()
    )


class GroupedGemmSingleLaunchRunner:
    """Host driver for the single-launch grouped GEMM (uniform per-group shape).

    Owns one :class:`KernelLauncher` over the batched universal-GEMM body.
    ``__call__`` issues a *single* device launch with ``grid.z = num_groups``;
    each group ``g`` is selected on-device by ``block_id_z == g`` and its
    matrices are addressed via the ``block_id_z * stride`` offset. The
    per-group tensors must be uniform shape and packed contiguously (the
    common batched-MoE layout); pass the contiguous base pointers + the shared
    ``(M, N, K)``.
    """

    def __init__(self, *, hsaco: bytes, spec: GroupedGemmSpec) -> None:
        self._spec = spec
        kernel_name = build_grouped_gemm_single_launch(spec).name
        self._launcher = KernelLauncher(
            hsaco=hsaco,
            kernel_name=kernel_name,
            signature=grouped_gemm_single_launch_signature(spec),
            cache_key=(kernel_name,),
        )
        self._block = (spec.block_size, 1, 1)
        self._tile_m = spec.tile.tile_m
        self._tile_n = spec.tile.tile_n

    def __call__(
        self,
        *,
        A_ptr: int,
        B_ptr: int,
        C_ptr: int,
        M: int,
        N: int,
        K: int,
        num_groups: int,
        stream: int = 0,
    ) -> LaunchSummary:
        cfg = LaunchConfig(
            stream=stream,
            grid=ceil_div_grid((N, self._tile_n), (M, self._tile_m), (num_groups, 1)),
            block=self._block,
        )
        summ = self._launcher(
            {
                "A": A_ptr,
                "B": B_ptr,
                "C": C_ptr,
                "M": M,
                "N": N,
                "K": K,
                "stride_a": M * K,
                "stride_b": N * K,
                "stride_c": M * N,
            },
            config=cfg,
        )
        return LaunchSummary(launches=summ.launches)


def grouped_gemm_problems(
    a_tensors: Iterable, b_tensors: Iterable, c_tensors: Iterable
) -> List[GroupedGemmProblem]:
    """Convenience: turn three lists of torch tensors into a problem list.

    ``a_tensors[g]`` is the ``(M_g, K)`` A matrix for group ``g``,
    ``b_tensors[g]`` is the ``(N_g, K)`` B matrix (note: B is stored
    transposed, matching the RCR convention), and ``c_tensors[g]`` is
    the ``(M_g, N)`` output matrix.
    """
    problems: List[GroupedGemmProblem] = []
    for a, bm, c in zip(a_tensors, b_tensors, c_tensors):
        M, K = a.shape[-2], a.shape[-1]
        N, K2 = bm.shape[-2], bm.shape[-1]
        if K != K2:
            raise ValueError(f"K mismatch: A K={K} vs B K={K2}")
        problems.append(
            GroupedGemmProblem(
                M=int(M),
                N=int(N),
                K=int(K),
                A_ptr=int(a.data_ptr()),
                B_ptr=int(bm.data_ptr()),
                C_ptr=int(c.data_ptr()),
            )
        )
    return problems

File: dnn-providers/hip-kernel-provider/rocke/platform/Cpp/instances/common/grouped_gemm.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/*
 * rocke/instance_grouped_gemm.c -- C99 port of
 * rocke/instances/common/grouped_gemm.py.
 *
 * GroupedGemmSpec is a thin wrapper around UniversalGemmSpec: every entry point
 * converts the grouped spec to a UniversalGemmSpec and delegates to the
 * already-ported gemm_universal machinery. No new IR is emitted, so the lowered
 * IR is byte-identical to the universal-GEMM body the Python delegates to.
 */
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_grouped_gemm.h"

#include "rocke/error_boundary.hpp" /* ckc::guard_builder boundary shim */
#include "rocke/helper_rocke.helpers.spec.h"
#include "rocke/ir_internal.h" /* rocke_i_set_err (sticky-error helper) */

/* ===================================================================== *
 *  rocke_grouped_gemm_spec_default
 *
 *  Python GroupedGemmSpec field defaults: wave_size=64, block_size=0,
 *  dtype="fp16". tile/trait carry their own dataclass defaults; trait reuses
 *  the universal-GEMM trait defaults (TraitSpec). The tile geometry is required
 *  and left zeroed for the caller to fill (matching the universal default).
 * ===================================================================== */
rocke_grouped_gemm_spec_t rocke_grouped_gemm_spec_default(void)
{
    rocke_grouped_gemm_spec_t s;
    rocke_gemm_universal_spec_t u = rocke_gemm_universal_spec_default();

    memset(&s, 0, sizeof(s));
    s.name = NULL;
    s.tile = u.tile; /* TileSpec defaults (warp_k=1, warp_tile 32/32/16) */
    s.trait = u.trait; /* TraitSpec defaults                               */
    s.wave_size = 64;
    s.block_size = 0;
    s.dtype = "fp16";
    return s;
}

/* ===================================================================== *
 *  rocke_grouped_gemm_spec_finalize
 *
 *  __post_init__ -> WarpTileBlockSizeMixin._init_block_size(): if block_size==0
 *  derive warp_m*warp_n*warp_k*wave_size. Idempotent.
 * ===================================================================== */
void rocke_grouped_gemm_spec_finalize(rocke_grouped_gemm_spec_t* spec)
{
    if(spec == NULL)
    {
        return;
    }
    spec->block_size = rocke_warp_tile_init_block_size(
        spec->block_size, spec->tile.warp_m, spec->tile.warp_n, spec->tile.warp_k, spec->wave_size);
}

/* ===================================================================== *
 *  rocke_grouped_gemm_data_spec
 *
 *  GroupedGemmSpec._data_spec():
 *      dt = "fp16" if self.dtype in ("f16", "fp16") else self.dtype
 *      return DataSpec(dtype_a=dt, dtype_b=dt, dtype_c=dt)
 *  The remaining DataSpec fields (dtype_acc, layout) take their universal-GEMM
 *  defaults ("fp32" / "RCR").
 * ===================================================================== */
rocke_gemm_data_spec_t rocke_grouped_gemm_data_spec(const rocke_grouped_gemm_spec_t* spec)
{
    rocke_gemm_universal_spec_t u = rocke_gemm_universal_spec_default();
    rocke_gemm_data_spec_t d = u.data; /* dtype_acc="fp32", layout="RCR" defaults */
    const char* dt;

    if(spec == NULL || spec->dtype == NULL)
    {
        dt = "fp16";
    }
    else if(strcmp(spec->dtype, "f16") == 0 || strcmp(spec->dtype, "fp16") == 0)
    {
        dt = "fp16";
    }
    else
    {
        dt = spec->dtype;
    }
    d.dtype_a = dt;
    d.dtype_b = dt;
    d.dtype_c = dt;
    return d;
}

/* ===================================================================== *
 *  rocke_grouped_gemm_to_universal_spec
 *
 *  GroupedGemmSpec.to_universal_spec():
 *      UniversalGemmSpec(name, tile, trait, data=_data_spec(),
 *                        wave_size, block_size, batched=False)
 *  The returned spec is finalized (block_size derived) so it is build-ready.
 * ===================================================================== */
rocke_gemm_universal_spec_t
    rocke_grouped_gemm_to_universal_spec(const rocke_grouped_gemm_spec_t* spec)
{
    rocke_gemm_universal_spec_t u = rocke_gemm_universal_spec_default();

    if(spec == NULL)
    {
        return u;
    }
    u.name = spec->name;
    u.tile = spec->tile;
    u.trait = spec->trait;
    u.data = rocke_grouped_gemm_data_spec(spec);
    u.wave_size = spec->wave_size;
    u.block_size = spec->block_size;
    u.batched = false;

    /* Python passes self.block_size (already derived by __post_init__). Mirror
     * that: ensure the universal spec's block_size matches the grouped spec's
     * finalized value. */
    rocke_gemm_universal_spec_finalize(&u);
    return u;
}

/* ===================================================================== *
 *  rocke_grouped_gemm_kernel_name
 *
 *  GroupedGemmSpec.kernel_name() == to_universal_spec().kernel_name().
 * ===================================================================== */
rocke_status_t
    rocke_grouped_gemm_kernel_name(const rocke_grouped_gemm_spec_t* spec, char* out, size_t out_cap)
{
    rocke_gemm_universal_spec_t u;

    if(spec == NULL || out == NULL || out_cap == 0)
    {
        return ROCKE_ERR_VALUE;
    }
    u = rocke_grouped_gemm_to_universal_spec(spec);
    return rocke_gemm_universal_kernel_name(&u, out, out_cap);
}

/* ===================================================================== *
 *  rocke_grouped_gemm_is_valid_spec
 *
 *  is_valid_spec(spec, arch): delegates to
 *  rocke.instances.common.gemm_universal.is_valid_spec(to_universal_spec()).
 * ===================================================================== */
bool rocke_grouped_gemm_is_valid_spec(const rocke_grouped_gemm_spec_t* spec,
                                      const char* arch,
                                      char* reason,
                                      size_t reason_cap)
{
    rocke_gemm_universal_spec_t u;

    if(spec == NULL)
    {
        if(reason != NULL && reason_cap > 0)
        {
            reason[0] = '\0';
        }
        return false;
    }
    if(arch == NULL)
    {
        arch = "gfx950";
    }
    u = rocke_grouped_gemm_to_universal_spec(spec);
    return rocke_gemm_universal_is_valid_spec(&u, arch, reason, reason_cap);
}

/* ===================================================================== *
 *  rocke_build_grouped_gemm
 *
 *  build_grouped_gemm(spec, arch):
 *      universal = spec.to_universal_spec()
 *      ok, why = is_valid_gemm_spec(universal, arch)
 *      if not ok: raise ValueError(...)
 *      return build_universal_gemm(universal, arch)
 *
 *  Python raises ValueError on reject; here we set the builder's sticky error
 *  and return NULL (so the caller's lower path surfaces the same message).
 * ===================================================================== */
rocke_kernel_def_t* rocke_build_grouped_gemm(rocke_ir_builder_t* b,
                                             const rocke_grouped_gemm_spec_t* spec,
                                             const char* arch)
{
    rocke_gemm_universal_spec_t u;
    char reason[ROCKE_ERR_MSG_CAP];

    if(b == NULL || spec == NULL)
    {
        return NULL;
    }
    if(arch == NULL)
    {
        arch = "gfx950";
    }

    u = rocke_grouped_gemm_to_universal_spec(spec);
    if(!rocke_gemm_universal_is_valid_spec(&u, arch, reason, sizeof(reason)))
    {
        return (rocke_kernel_def_t*)rocke_i_set_err(
            b, ROCKE_ERR_VALUE, "invalid grouped_gemm spec for %s: %s", arch, reason);
    }
    return rocke_build_universal_gemm(b, &u, arch);
}

/* ===================================================================== *
 *  rocke_build_grouped_gemm_new -- init builder with spec.kernel_name() + build.
 * ===================================================================== */
rocke_kernel_def_t* rocke_build_grouped_gemm_new(rocke_ir_builder_t* b,
                                                 const rocke_grouped_gemm_spec_t* spec,
                                                 const char* arch)
{
    return ckc::guard_builder(b, [&]() -> rocke_kernel_def_t* {
        char name[256];

        if(b == NULL || spec == NULL)
        {
            return NULL;
        }
        if(rocke_grouped_gemm_kernel_name(spec, name, sizeof(name)) != ROCKE_OK)
        {
            return NULL;
        }
        if(rocke_ir_builder_init(b, name) != ROCKE_OK)
        {
            return NULL;
        }
        return rocke_build_grouped_gemm(b, spec, arch);
    });
}

/* ===================================================================== *
 *  rocke_build_grouped_gemm_single_launch
 *
 *  build_grouped_gemm_single_launch(spec, arch):
 *      base_spec = spec.to_universal_spec()
 *      batched_spec = UniversalGemmSpec(name=base_spec.name + "_single_launch",
 *                                       tile, trait, data, wave_size,
 *                                       block_size, batched=True)
 *      ok, why = is_valid_gemm_spec(batched_spec, arch)
 *      if not ok: raise ValueError(...)
 *      return build_universal_gemm(batched_spec, arch)
 * ===================================================================== */
rocke_kernel_def_t* rocke_build_grouped_gemm_single_launch(rocke_ir_builder_t* b,
                                                           const rocke_grouped_gemm_spec_t* spec,
                                                           const char* arch)
{
    rocke_gemm_universal_spec_t u;
    char reason[ROCKE_ERR_MSG_CAP];
    char name[256];
    const char* base_name;
    size_t blen, slen;

    if(b == NULL || spec == NULL)
    {
        return NULL;
    }
    if(arch == NULL)
    {
        arch = "gfx950";
    }

    u = rocke_grouped_gemm_to_universal_spec(spec);

    /* name = base_spec.name + "_single_launch" */
    base_name = (u.name != NULL) ? u.name : "";
    blen = strlen(base_name);
    slen = strlen("_single_launch");
    if(blen + slen >= sizeof(name))
    {
        return (rocke_kernel_def_t*)rocke_i_set_err(
            b, ROCKE_ERR_VALUE, "grouped_gemm_single_launch: name too long");
    }
    memcpy(name, base_name, blen);
    memcpy(name + blen, "_single_launch", slen + 1);
    u.name = name;
    u.batched = true;

    if(!rocke_gemm_universal_is_valid_spec(&u, arch, reason, sizeof(reason)))
    {
        return (rocke_kernel_def_t*)rocke_i_set_err(
            b, ROCKE_ERR_VALUE, "invalid grouped_gemm_single_launch spec for %s: %s", arch, reason);
    }
    return rocke_build_universal_gemm(b, &u, arch);
}

/* ===================================================================== *
 *  rocke_build_grouped_gemm_single_launch_new
 * ===================================================================== */
rocke_kernel_def_t* rocke_build_grouped_gemm_single_launch_new(
    rocke_ir_builder_t* b, const rocke_grouped_gemm_spec_t* spec, const char* arch)
{
    return ckc::guard_builder(b, [&]() -> rocke_kernel_def_t* {
        rocke_gemm_universal_spec_t u;
        char name[256];
        const char* base_name;
        size_t blen, slen;

        if(b == NULL || spec == NULL)
        {
            return NULL;
        }

        /* Compute the single-launch kernel name = <base kernel_name>_single_launch.
         * Python: build_grouped_gemm_single_launch(spec).name on the renamed spec.
         * The kernel name is derived from the (renamed) UniversalGemmSpec. */
        u = rocke_grouped_gemm_to_universal_spec(spec);
        base_name = (u.name != NULL) ? u.name : "";
        blen = strlen(base_name);
        slen = strlen("_single_launch");
        if(blen + slen >= sizeof(name))
        {
            return NULL;
        }
        memcpy(name, base_name, blen);
        memcpy(name + blen, "_single_launch", slen + 1);
        u.name = name;
        u.batched = true;

        {
            char kname[256];
            if(rocke_gemm_universal_kernel_name(&u, kname, sizeof(kname)) != ROCKE_OK)
            {
                return NULL;
            }
            if(rocke_ir_builder_init(b, kname) != ROCKE_OK)
            {
                return NULL;
            }
        }
        return rocke_build_grouped_gemm_single_launch(b, spec, arch);
    });
}

/* ===================================================================== *
 *  signature helpers
 *
 *  ptr_dt = spec.dtype if spec.dtype in ("f16","fp16","bf16") else "f16".
 * ===================================================================== */
static const char* grouped_gemm_ptr_dt(const rocke_grouped_gemm_spec_t* spec)
{
    const char* dt = (spec != NULL) ? spec->dtype : NULL;
    if(dt != NULL && (strcmp(dt, "f16") == 0 || strcmp(dt, "fp16") == 0 || strcmp(dt, "bf16") == 0))
    {
        return dt;
    }
    return "f16";
}

/* grouped_gemm_signature(spec): (A,B,C ptr ; M,N,K i32). */
rocke_status_t rocke_grouped_gemm_signature(const rocke_grouped_gemm_spec_t* spec,
                                            rocke_arena_t* arena,
                                            const rocke_sig_entry_t** out_items,
                                            size_t* out_count)
{
    rocke_signature_builder_t sb;
    const char* ptr_dt;
    rocke_status_t st;

    if(spec == NULL || arena == NULL)
    {
        return ROCKE_ERR_VALUE;
    }
    st = rocke_signature_builder_init(&sb, arena);
    if(st != ROCKE_OK)
    {
        return st;
    }
    ptr_dt = grouped_gemm_ptr_dt(spec);

    rocke_signature_builder_ptr(&sb, "A", ptr_dt, NULL);
    rocke_signature_builder_ptr(&sb, "B", ptr_dt, NULL);
    rocke_signature_builder_ptr(&sb, "C", ptr_dt, NULL);
    rocke_signature_builder_scalar(&sb, "M", "i32");
    rocke_signature_builder_scalar(&sb, "N", "i32");
    rocke_signature_builder_scalar(&sb, "K", "i32");

    return rocke_signature_builder_build(&sb, out_items, out_count);
}

/* grouped_gemm_single_launch_signature(spec):
 *   (A,B,C ptr ; M,N,K i32 ; stride_a,stride_b,stride_c i32). */
rocke_status_t rocke_grouped_gemm_single_launch_signature(const rocke_grouped_gemm_spec_t* spec,
                                                          rocke_arena_t* arena,
                                                          const rocke_sig_entry_t** out_items,
                                                          size_t* out_count)
{
    rocke_signature_builder_t sb;
    const char* ptr_dt;
    rocke_status_t st;

    if(spec == NULL || arena == NULL)
    {
        return ROCKE_ERR_VALUE;
    }
    st = rocke_signature_builder_init(&sb, arena);
    if(st != ROCKE_OK)
    {
        return st;
    }
    ptr_dt = grouped_gemm_ptr_dt(spec);

    rocke_signature_builder_ptr(&sb, "A", ptr_dt, NULL);
    rocke_signature_builder_ptr(&sb, "B", ptr_dt, NULL);
    rocke_signature_builder_ptr(&sb, "C", ptr_dt, NULL);
    rocke_signature_builder_scalar(&sb, "M", "i32");
    rocke_signature_builder_scalar(&sb, "N", "i32");
    rocke_signature_builder_scalar(&sb, "K", "i32");
    rocke_signature_builder_scalar(&sb, "stride_a", "i32");
    rocke_signature_builder_scalar(&sb, "stride_b", "i32");
    rocke_signature_builder_scalar(&sb, "stride_c", "i32");

    return rocke_signature_builder_build(&sb, out_items, out_count);
}

/* ===================================================================== *
 *  rocke_grouped_gemm_lower_to_llvm -- build the per-group base kernel + lower.
 *  Owns and frees its own IRBuilder (mirrors rocke_gemm_universal_lower_to_llvm).
 * ===================================================================== */
rocke_status_t rocke_grouped_gemm_lower_to_llvm(const rocke_grouped_gemm_spec_t* spec,
                                                const char* arch,
                                                rocke_llvm_flavor_t flavor,
                                                char** out_ll,
                                                char* err,
                                                size_t err_cap)
{
    rocke_ir_builder_t b;
    rocke_kernel_def_t* kernel;
    rocke_status_t st;

    if(out_ll != NULL)
    {
        *out_ll = NULL;
    }
    if(spec == NULL || out_ll == NULL)
    {
        if(err != NULL && err_cap > 0)
        {
            const char* m = "grouped_gemm lower_to_llvm: null spec/out";
            size_t n = strlen(m);
            if(n >= err_cap)
            {
                n = err_cap - 1;
            }
            memcpy(err, m, n);
            err[n] = '\0';
        }
        return ROCKE_ERR_VALUE;
    }
    if(arch == NULL)
    {
        arch = "gfx950";
    }

    kernel = rocke_build_grouped_gemm_new(&b, spec, arch);
    if(kernel == NULL)
    {
        st = rocke_ir_builder_status(&b);
        if(err != NULL && err_cap > 0)
        {
            const char* m = rocke_ir_builder_error(&b);
            size_t n;
            if(m == NULL)
            {
                m = "build_grouped_gemm failed";
            }
            n = strlen(m);
            if(n >= err_cap)
            {
                n = err_cap - 1;
            }
            memcpy(err, m, n);
            err[n] = '\0';
        }
        rocke_ir_builder_free(&b);
        return (st == ROCKE_OK) ? ROCKE_ERR_VALUE : st;
    }

    st = rocke_lower_kernel_to_llvm_ex(kernel, flavor, arch, out_ll, err, err_cap);
    rocke_ir_builder_free(&b);
    return st;
}
